File: Learning/src/Questions/DP/EditDistance.java
package Questions.DP;

import java.util.Arrays;

public class EditDistance {
    // We are given two strings ‘S1’ and ‘S2’. We need to convert S1 to S2. The
    // following three operations are allowed:

    // Deletion of a character.
    // Replacement of a character with another one.
    // Insertion of a character.
    // We have to return the minimum number of operations required to convert S1 to
    // S2 as our answer.

    // Function to calculate the minimum edit distance between two strings
    static int editDistanceUtil(String S1, String S2, int i, int j, int[][] dp) {
        // Base cases
        if (i < 0)
            return j + 1;
        if (j < 0)
            return i + 1;

        // If the result is already computed, return it
        if (dp[i][j] != -1)
            return dp[i][j];

        // If the characters at the current positions match, no edit is needed
        if (S1.charAt(i) == S2.charAt(j))
            return dp[i][j] = editDistanceUtil(S1, S2, i - 1, j - 1, dp);

        // Minimum of three choices:
        // 1. Replace the character in S1 with the character in S2.
        // 2. Delete the character in S1.
        // 3. Insert the character from S2 into S1.
        else
            return dp[i][j] = 1 + Math.min(editDistanceUtil(S1, S2, i - 1, j - 1, dp),
                    Math.min(editDistanceUtil(S1, S2, i - 1, j, dp), editDistanceUtil(S1, S2, i, j - 1, dp)));
    }

    static int editDistanceRecursive(String S1, String S2) {
        int n = S1.length();
        int m = S2.length();

        int[][] dp = new int[n][m];
        for (int row[] : dp)
            Arrays.fill(row, -1);

        // Call the recursive helper function
        return editDistanceUtil(S1, S2, n - 1, m - 1, dp);
    }

    // Time Complexity: O(N*M)

    // Reason: There are N*M states therefore at max ‘N*M’ new problems will be
    // solved.

    // Space Complexity: O(N*M) + O(N+M)

    // Reason: We are using a recursion stack space(O(N+M)) and a 2D array (
    // O(N*M)).

    // Function to calculate the minimum edit distance between two strings
    static int editDistanceTabulation(String S1, String S2) {
        int n = S1.length();
        int m = S2.length();

        // Create a 2D array to store the minimum edit distances
        // Each cell (i, j) in the array represents the minimum edit distance between
        // the
        // S1[0..i-1] and S2[0..j-1]
        int[][] dp = new int[n + 1][m + 1];

        // If one of the strings is empty, the minimum edit distance is the length of
        // the other string
        for (int i = 0; i <= n; i++) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= m; j++) {
            dp[0][j] = j;
        }

        // Fill the dp array using a bottom-up approach
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (S1.charAt(i - 1) == S2.charAt(j - 1)) {
                    // If the characters match, no edit is needed, so take the value from the
                    // diagonal.
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    // If the characters don't match, take the minimum of three possibilities:
                    // 1. Replace the character in S1 with the character in S2 (diagonal).
                    // 2. Delete the character in S1 (left).
                    // 3. Insert the character from S2 into S1 (up).
                    dp[i][j] = 1 + Math.min(dp[i - 1][j - 1], Math.min(dp[i - 1][j], dp[i][j - 1]));
                }
            }
        }

        return dp[n][m];
    }
    // Time Complexity: O(N*M)

    // Reason: There are two nested loops

    // Space Complexity: O(N*M)

    // Reason: We are using an external array of size ‘N*M’. Stack Space is
    // eliminated.

    // Function to calculate the minimum edit distance between two strings
    static int editDistance(String S1, String S2) {
        int n = S1.length();
        int m = S2.length();

        // Create two arrays to store the previous and current rows of minimum edit
        // distances
        int[] prev = new int[m + 1];
        int[] cur = new int[m + 1];

        // Initialize the first row with their respective indices
        for (int j = 0; j <= m; j++) {
            prev[j] = j;
        }

        // Fill the cur array using a bottom-up approach
        for (int i = 1; i <= n; i++) {
            cur[0] = i;
            for (int j = 1; j <= m; j++) {
                if (S1.charAt(i - 1) == S2.charAt(j - 1)) {
                    // If the characters match, no edit is needed, so take the value from the
                    // diagonal.
                    cur[j] = prev[j - 1];
                } else {
                    // If the characters don't match, take the minimum of three possibilities:
                    // 1. Replace the character in S1 with the character in S2 (diagonal).
                    // 2. Delete the character in S1 (left).
                    // 3. Insert the character from S2 into S1 (up).
                    cur[j] = 1 + Math.min(prev[j - 1], Math.min(prev[j], cur[j - 1]));
                }
            }
            // Update prev array to store the current values
            prev = cur.clone();
        }

        return cur[m];
    }

    public static void main(String[] args) {
        String S1 = "sunday";
        String S2 = "saturday";
        System.out.println(editDistanceRecursive(S1, S2)); // 3
        System.out.println(editDistanceTabulation(S1, S2)); // 3
        System.out.println(editDistance(S1, S2)); // 3
    }
    // Time Complexity: O(N*M)

    // Reason: There are two nested loops.

    // Space Complexity: O(M)

    // Reason: We are using an external array of size ‘M+1’ to store two rows.
}

File: DSA/Python/Questions/DP/edit_distance.py
"""
# Question: Edit Distance
# Link: https://leetcode.com/problems/edit-distance/

# Problem Statement:
# Given two strings str1 and str2, return the minimum number of operations
# required to convert str1 to str2. The allowed operations are:
# 1. Insert a character
# 2. Delete a character
# 3. Replace a character

# Example:
# Input: str1 = "horse", str2 = "ros"
# Output: 3
# Explanation: 
# horse -> rorse (replace 'h' with 'r')
# rorse -> rose (remove 'r')
# rose -> ros (remove 'e')

# Key Insights:
# - Each position can have three possible operations
# - Need to track minimum operations at each step
# - Characters matching means no operation needed
"""

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. Compare characters from end of both strings
2. If characters match, no operation needed
3. If different, try all three operations:
   - Replace current character
   - Delete character from str1
   - Insert character from str2
4. Use memoization to store results

Time Complexity: O(n*m)
Space Complexity: O(n*m) + O(n+m) recursion stack
"""


def edit_distance_recursive(str1: str, str2: str) -> int:
    n, m = len(str1), len(str2)
    # Initialize dp array for memoization
    dp = [[-1] * m for _ in range(n)]

    def solve(i: int, j: int) -> int:
        # Base cases: if either string is exhausted
        if i < 0:
            return j + 1
        if j < 0:
            return i + 1

        # Return memoized result if available
        if dp[i][j] != -1:
            return dp[i][j]

        # If characters match, no operation needed
        if str1[i] == str2[j]:
            dp[i][j] = solve(i - 1, j - 1)
            return dp[i][j]

        # Try all three operations and take minimum
        replace = solve(i - 1, j - 1)
        delete = solve(i - 1, j)
        insert = solve(i, j - 1)

        # Store minimum operations needed
        dp[i][j] = 1 + min(replace, delete, insert)
        return dp[i][j]

    return solve(n - 1, m - 1)


# 2. Tabulation Solution
"""
Algorithm:
1. Create 2D dp array of size (n+1)*(m+1)
2. Initialize base cases for empty strings
3. Fill dp table using bottom-up approach
4. For each position, calculate minimum operations

Time Complexity: O(n*m)
Space Complexity: O(n*m)
"""


def edit_distance_tabulation(str1: str, str2: str) -> int:
    n, m = len(str1), len(str2)
    # Create dp array with base cases
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    # Initialize first row and column
    for i in range(n + 1):
        dp[i][0] = i
    for j in range(m + 1):
        dp[0][j] = j

    # Fill dp table
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            # If characters match
            if str1[i - 1] == str2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                # Take minimum of three operations
                dp[i][j] = 1 + min(
                    dp[i - 1][j - 1],  # replace
                    dp[i - 1][j],  # delete
                    dp[i][j - 1],  # insert
                )

    return dp[n][m]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use two arrays instead of 2D array
2. Previous row and current row sufficient
3. Update arrays in alternating fashion

Time Complexity: O(n*m)
Space Complexity: O(m)
"""


def edit_distance_optimized(str1: str, str2: str) -> int:
    n, m = len(str1), len(str2)
    # Initialize previous row
    prev = [j for j in range(m + 1)]

    # Process each character of str1
    for i in range(1, n + 1):
        # Current row array
        curr = [0] * (m + 1)
        curr[0] = i  # Base case

        # Fill current row
        for j in range(1, m + 1):
            if str1[i - 1] == str2[j - 1]:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(
                    prev[j - 1], prev[j], curr[j - 1]  # replace  # delete  # insert
                )

        # Update previous row
        prev = curr[:]

    return prev[m]


def main():
    # Test cases
    test_cases = [
        ("horse", "ros"),  # Regular case
        ("intention", "execution"),  # Multiple operations
        ("", "abc"),  # Empty string
        ("abc", ""),  # Empty target
        ("sunday", "saturday"),  # Complex transformations
    ]

    for i, (str1, str2) in enumerate(test_cases):
        print(f"\nTest Case {i + 1}: str1 = '{str1}', str2 = '{str2}'")
        print(f"Recursive Solution: {edit_distance_recursive(str1, str2)}")
        print(f"Tabulation Solution: {edit_distance_tabulation(str1, str2)}")
        print(f"Optimized Solution: {edit_distance_optimized(str1, str2)}")


if __name__ == "__main__":
    main()
